### PR TITLE
Introduce min ballot duration in VotingToManageEmissionFunds contract

### PR DESCRIPTION
**Pull request by maxaleks**
https://github.com/poanetwork/poa-network-consensus-contracts/issues/190

### CODE REVIEW

File: contracts/VotingToManageEmissionFunds.sol [hunk 1]
```diff
@@ -159,14 +165,16 @@ contract VotingToManageEmissionFunds is VotingTo {
         uint256 _emissionReleaseTime, // unix timestamp
         uint256 _emissionReleaseThreshold, // seconds
         uint256 _distributionThreshold, // seconds
-        address _emissionFunds
+        address _emissionFunds,
+        uint256 _minBallotDuration
```
Comment (varasev) [inline contracts/VotingToManageEmissionFunds.sol]: Since we add this param here, let's fix `migrations/2_deploy_contract.js` accordingly.

File: contracts/VotingToManageEmissionFunds.sol [hunk 2]
```diff
@@ -159,14 +165,16 @@ contract VotingToManageEmissionFunds is VotingTo {
...
     ) public onlyOwner {
         require(!initDisabled());
```
Comment (varasev) [inline contracts/VotingToManageEmissionFunds.sol]: Seems we don't need this line here anymore because it is already present in `VotingTo._init` function. The same goes for the line `boolStorage[INIT_DISABLED] = true;` below.

File: test/voting_to_manage_emission_funds_test.js [hunk 3]
```diff
@@ -241,6 +247,13 @@ contract('VotingToManageEmissionFunds [all features]', function (accounts) {
         VOTING_START_DATE, VOTING_END_DATE, accounts[5], "memo", {from: votingKey}
       ).should.be.rejectedWith(ERROR_MSG);
     });
+    it('ballot duration must be greater than min ballot duration', async () => {
+      VOTING_START_DATE = moment.utc().add(31 * 60, 'seconds').unix();
+      VOTING_END_DATE = moment.utc().add(31 * 60 + minBallotDuration - 1, 'seconds').unix();
+      await voting.createBallot(
+        VOTING_START_DATE, VOTING_END_DATE, accounts[5], "memo", {from: votingKey}
+      ).should.be.rejectedWith(ERROR_MSG);
```
Comment (varasev) [inline test/voting_to_manage_emission_funds_test.js]: Please add here one more test (right after this line) which would show that the `createBallot` is called successfully for `VOTING_END_DATE = moment.utc().add(31 * 60 + minBallotDuration + 1, 'seconds').unix();`.

**Review by varasev — CHANGES_REQUESTED** [review]
Please also review `require` statements related to checking the `_startTime` and `_endTime` parameters in the `VotingToManageEmissionFunds.createBallot` function. Since now we use the `onlyValidTime` modifier inside the `VotingTo._createBallot` internal function, those `require` statements in `VotingToManageEmissionFunds.createBallot` look redundant.

File: contracts/VotingToManageEmissionFunds.sol [hunk 4]
```diff
@@ -159,20 +163,20 @@ contract VotingToManageEmissionFunds is VotingTo {
         uint256 _emissionReleaseTime, // unix timestamp
         uint256 _emissionReleaseThreshold, // seconds
         uint256 _distributionThreshold, // seconds
-        address _emissionFunds
+        address _emissionFunds,
+        uint256 _minBallotDuration
```
Comment (varasev) [inline contracts/VotingToManageEmissionFunds.sol]: I forgot one more thing: `_minBallotDuration` here must be less than the `_distributionThreshold`. Otherwise, a ballot cannot be created.

Let's also note in a comment that `_minBallotDuration` is defined in seconds and make this parameter fourth (not fifth, for security reasons, because the length of the `address _emissionFunds` parameter can be a float, with leading zeros).

**Comment by varasev** [comment]
Let's also add `canBeFinalizedNow` into this condition: https://github.com/poanetwork/poa-network-consensus-contracts/blob/f7a374573200e5b002338036cc17f4fb2ec85d14/contracts/VotingToManageEmissionFunds.sol#L223 and add an additional unit tests to `test/voting_to_manage_emission_funds_*.js` for the `VotingToManageEmissionFunds.finalize`, `vote`, and `init` functions to cover the added code by tests.

File: test/voting_to_manage_emission_funds_test.js [hunk 5]
```diff
@@ -545,7 +545,8 @@ contract('VotingToManageEmissionFunds [all features]', function (accounts) {
       const emissionFundsBalanceOld = await web3.eth.getBalance(emissionFunds.address);
       const receiverBalanceOld = await web3.eth.getBalance(receiver);
 
-      await voting.setTime(VOTING_START_DATE);
+      const time = VOTING_START_DATE + minBallotDuration + 1;
+      await voting.setTime(time);
```
Comment (varasev) [inline test/voting_to_manage_emission_funds_test.js]: Why did you change this to `VOTING_START_DATE + minBallotDuration + 1`? The vote must pass right beginning from `VOTING_START_DATE`. The voter shouldn't wait until `minBallotDuration` is expired.

Comment (varasev) [reply test/voting_to_manage_emission_funds_test.js]: Is this because of `ballotInfo[4].should.be.equal(true); // isFinalized` ?

Comment (maxaleks) [reply test/voting_to_manage_emission_funds_test.js]: > Is this because of `ballotInfo[4].should.be.equal(true); // isFinalized` ?

Yes, it is

**Comment by varasev** [comment]
Now, we need to try to launch [`poa-test-setup`](https://github.com/poanetwork/poa-test-setup) with these changed contracts: if it goes well, we will merge this PR. Please prepare a separate branch for `poa-test-setup` which would use `min-ballot-duration-for-manage-emission-funds` branch from your repo (instead of `master` of `poa-network-consensus-contracts` repo), and try to launch all the steps of `poa-test-setup`.

**Review by varasev — APPROVED** [review]
